ibis/backends/mssql/compiler.py

```diff
@@ -1,39 +1,599 @@
+import ibis.expr.datatypes as dt
+import ibis.expr.operations as ops
+from ibis.backends.base.sqlglot.compiler import NULL, STAR, SQLGlotCompiler, paren
+from ibis.backends.base.sqlglot.datatypes import MSSQLType
+from ibis.backends.base.sqlglot.rewrites import (
+    rewrite_first_to_first_value,
+    rewrite_last_to_last_value,
+)
+from ibis.common.deferred import var
+from ibis.common.patterns import replace
+from ibis.expr.rewrites import p, rewrite_sample
 
-class MsSqlExprTranslator(AlchemyExprTranslator):
-    _registry = operation_registry
-    _rewrites = AlchemyExprTranslator._rewrites.copy()
-    _bool_aggs_need_cast_to_int32 = True
+TSQL.Generator.TRANSFORMS |= {
+    sge.ApproxDistinct: rename_func("approx_count_distinct"),
+    sge.Stddev: rename_func("stdevp"),
+    sge.StddevPop: rename_func("stdevp"),
+    sge.StddevSamp: rename_func("stdev"),
+    sge.Variance: rename_func("var"),
+    sge.VariancePop: rename_func("varp"),
+    sge.Ceil: rename_func("ceiling"),
+    sge.Trim: lambda _, e: f"TRIM({e.this.sql('tsql')})",
+}
 
-    _timestamp_type = DATETIME2
-    _integer_to_timestamp = staticmethod(_timestamp_from_unix)
+x = var("x")
+y = var("y")
+start = var("start")
+end = var("end")
 
-    native_json_type = False
 
-    _forbids_frame_clause = AlchemyExprTranslator._forbids_frame_clause + (
-        ops.Lag,
-        ops.Lead,
+@replace(
+    p.WindowFunction(
+        p.RowNumber | p.NTile | p.Lag | p.Lead | p.PercentRank | p.CumeDist, y
     )
-    _require_order_by = AlchemyExprTranslator._require_order_by + (ops.Reduction,)
-    _dialect_name = "mssql"
+)
+def exclude_unsupported_window_frame_from_ops(_, y):
+    return _.copy(frame=y.copy(start=None, end=0))
+
+
+@replace(p.Not(x))
+def replace_not_with_ifelse(_, x):
+    assert _.dtype.is_boolean()
+    return ops.Cast(ops.IfElse(x, 0, 1), _.dtype)
+
+
+@replace(p.Not(x))
+def replace_not_with_equals(_, x):
+    return ops.Equals(x, 0)
+
+
+def any_all_to_max_min(_, cls):
+    assert _.arg.dtype.is_boolean()
+    return ops.Cast(cls(ops.IfElse(_.arg, 1, 0), where=_.where), _.arg.dtype)
+
+
+@replace(p.Any)
+def any_to_max(_):
+    return any_all_to_max_min(_, ops.Max)
+
+
+@replace(p.All)
+def all_to_min(_):
+    return any_all_to_max_min(_, ops.Min)
+
+
+@replace(p.Between)
+def between_to_integer(_):
+    assert _.dtype.is_boolean()
+    return ops.Cast(ops.IfElse(_, 1, 0), _.dtype)
+
+
+project_only_replacements = reduce(
+    operator.or_,
+    (
+        replace_not_with_ifelse,
+        between_to_integer,
+        any_to_max,
+        all_to_min,
+        rewrite_first_to_first_value,
+        rewrite_last_to_last_value,
+        exclude_unsupported_window_frame_from_ops,
+    ),
+)
+
+
+@replace(p.DummyTable)
+def replace_dummy_booleans(_, **kwargs):
+    values = _.values
+    subs = {}
+    for value in values.values():
+        subs[value] = value.replace(project_only_replacements, filter=ops.Value)
+    return _.replace(subs)
```

Comment:
While this replacement is correct it is traversing `_` twice. I would suggest to simply:

```py
values = {k: v.replace(project_only_replacements, filter=ops.Value for k, v in _.values.items()}
return _.copy(values=values)
```